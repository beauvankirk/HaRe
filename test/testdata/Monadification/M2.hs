module M1 where
import Data.Char

data Expr = Var Char
          | N Int
          | Add Expr Expr
          | Sub Expr Expr
          | Assign Char Expr
          deriving Show

type Env = [(Char,Int)]

eval :: Expr -> Env -> (Int,Env)

eval (Var v) env = (head [val | (x,val) <- env, x==v], env)

eval (N n) env = (n,env)

eval (Add e1 e2) env = let (v1,env1) = eval e1 env
                           (v2,env2) = eval e2 env1 in
  (v1+v2,env2)
eval (Sub e1 e2) env = let (v1,env1) = eval e1 env
                           (v2,env2) = eval e2 env1 in
  (v1-v2,env2)
eval (Assign x e) env = let (v,env1) = eval e env in
  (v, (x,v):env1)
