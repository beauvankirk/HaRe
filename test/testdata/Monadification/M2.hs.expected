module M1 where
import Data.Char

data Expr = Var Char
          | N Int
          | Add Expr Expr
          | Sub Expr Expr
          | Assign Char Expr
          deriving Show

type Env = [(Char,Int)]

eval :: Monad m => Expr -> Env -> m (Int,Env)

eval (Var v) env = return (head [val | (x,val) <- env, x==v], env)

eval (N n) env = return (n,env)

eval (Add e1 e2) env = eval e1 env >>= (\(v1,env1) -> eval e2 env1 >>= (\(v2,env2) -> return (v1+v2,env2)))
eval (Sub e1 e2) env = eval e1 env >>= (\(v1,env1) -> eval e2 env1 >>= (\(v2,env2) -> return (v1-v2,env2)))
eval (Assign x e) env = eval e env >>= (\(v,env1) -> return (v, (x,v):env1))
